src/charm.py

```diff
@@ -90,23 +91,36 @@ def _on_config_changed(self, _) -> None:
         # Set the cluster name in the peer relation databag if it is not already set
         peer_data = self._peers.data[self.app]
 
-        if not peer_data.get("cluster_name"):
-            peer_data["cluster_name"] = self.config.get("cluster_name") or generate_random_hash()
+        if not peer_data.get("cluster-name"):
+            peer_data["cluster-name"] = (
+                self.config.get("cluster-name") or f"cluster_{generate_random_hash()}"
+            )
 
     def _on_start(self, _) -> None:
         """Handle the start event."""
+        # Configure MySQL users and the instance for use in an InnoDB cluster
         try:
-            # TODO: add logic to determine how to add non-leader instances to the cluster
-            if self.unit.is_leader():
-                self._mysql.configure_mysql_users()
-                self._mysql.configure_instance()
+            self._mysql.configure_mysql_users()
+            self._mysql.configure_instance()
         except MySQLConfigureMySQLUsersError:
             self.unit.status = BlockedStatus("Failed to initialize MySQL users")
             return
         except MySQLConfigureInstanceError:
             self.unit.status = BlockedStatus("Failed to configure instance for InnoDB")
             return
 
+        # Create the cluster on the juju leader unit
+        if not self.unit.is_leader():
+            self.unit.status = ActiveStatus()
```

Comment:
Common lifecycle question here:
You're setting non leader unit as Active, but the unit at this point is not a cluster member.
Wouldn't be more correct to set it to WaitingStatus (e.g. "waiting for peer relation join")?

And then, there's another question: since the leader unit is the one which join the incoming non-leader unit to the cluster, how to update it status from waiting->active, after a successful add_instance that was done by the leader?

Reply:
Discussed offline with Paulo: non-leader units should be moved into `WaitingStatus`. And then, in the `update-status` event handler, we'd move the unit to `ActiveStatus` after confirming that it is part of the InnoDB cluster.

We will implement this in a future PR - added a TODO item in fc409d658a9d226f29dd0b36a2723615e638dd98